packages/lesswrong/server/migrations/2019-11-24-editableLatestRevision1.js

```diff
@@ -0,0 +1,107 @@
+import { getCollection } from 'meteor/vulcan:core'
+import { editableCollections, editableCollectionsFields } from '../../lib/editor/make_editable';
+import { Revisions } from '../../lib/index';
+
+// The upgrade procedure here is:
+//  1. Attach an instance to the database and run editableAddLatestRevisionField.
+//     This creates a [fieldName]_latest field for all documents with
+//     content-editable fields.
+//  2. Update the webserver. This makes the [fieldName]_latest field be the one
+//     that matters, rather than [fieldName]. But any edits that happen on the
+//     old server version will have [fieldName]_latest out of date.
+//  3. Run editableAddLatestRevisionField again to repair any documents that were
+//     edited during the upgrade
+//  4. Run editableDropDenormalizedField to drop the now-unused denormalized
+//     content fields, for smaller tables and faster database operations.
+
+registerMigration({
+  name: "editableAddLatestRevisionField",
+  idempotent: true,
+  action: async () => {
+    for (let collectionName of editableCollections)
+    for (let fieldName of editableCollectionsFields[collectionName])
+    {
+      const collection = getCollection(collectionName);
+      // eslint-disable-next-line no-console
+      console.log(`Updating ${collectionName}.${fieldName}_latest`);
+      await forEachDocumentBatchInCollection({
+        collection: collection,
+        filter: {
+          [fieldName]: {$exists: true},
+          [`${fieldName}_latest`]: {$exists: false},
+        },
+        batchSize: 1000,
+        callback: async (documents) => {
+          const updates = [];
+          await Promise.all(
+            documents.map(async doc => {
+              if (doc[fieldName]) {
+                const latestRev = await Revisions.findOne({
+                  documentId: doc._id,
+                  fieldName: fieldName,
+                  version: doc[fieldName].version,
+                }, {}, {_id: 1});
+                if (latestRev) {
+                  updates.push({
+                    updateOne: {
+                      filter: { _id: doc._id },
+                      update: {
+                        $set: {
+                          [`${fieldName}_latest`]: latestRev._id
+                        }
+                      }
+                    }
+                  });
+                } else {
+                  // eslint-disable-next-line no-console
+                  console.log(`Warning: document is missing its corresponding revision object`);
+                }
+              }
+            })
+          );
+          
+          if (updates.length > 0) {
+            // eslint-disable-next-line no-console
+            console.log(`Updating ${updates.length} documents in ${collectionName}`);
+            await collection.rawCollection().bulkWrite(updates, { ordered: false });
+          }
+        }
+      });
+    }
+  }
+});
+
+/*registerMigration({
```

Comment:
Presumably this shouldn't be commented out?

Reply:
This migration doesn't make sense to run until after the first migration is run *and* the next PR after this one is merged, so it's temporarily a footgun. Will uncomment later.